File: frontend/src/client-host.ts
// The renderer half of the CLIENT HOST (nocx-uo1k6, design D3).
//
// The coordinator is a daemon with no window: a file picker, a browser open,
// a desktop banner, a dock badge and a window raise are things only this
// client can do. So the coordinator asks (host.request) and this module
// answers (host.resolved), performing the effect through the Wails bindings
// on the way.
//
// THE RENDERER DECIDES NOTHING (AD-3). Whether the URL may be opened, whether
// a second picker may stack, which pane a click focuses -- all of it is
// settled on the coordinator's side before the ask arrives. This module
// performs and reports; the one judgement it makes is "this client has no
// native host at all", which is a fact about the environment, not a policy.
//
// Every path answers. A request dropped silently is a coordinator task
// waiting on a person who was never asked, so an unknown capability, a
// missing runtime and a thrown binding all resolve -- failed, with a
// sentence.

import { Events } from '@wailsio/runtime'
import {
  HostBadge,
  HostBanner,
  HostBounce,
  HostFocusWindow,
  HostOpenDirectory,
  HostOpenFile,
  HostOpenUrl,
} from '../bindings/github.com/shady2k/nocx/wailsapp'
import { bindingReachable } from './wails-runtime'
import type { Dispatcher } from './dispatcher'
import type { HostAttentionActivated } from './generated/host.attentionActivated'
import type { HostRequest } from './generated/host.request'
import type { HostResolved } from './generated/host.resolved'

/** The native effects this client can perform, as a seam. The default binds
 *  the generated Wails bindings; a test binds doubles and needs no runtime. */
export interface HostBindings {
  openFile(): Promise<string>
  openDirectory(): Promise<string>
  openUrl(url: string): Promise<void>
  banner(title: string, body: string, sessionId: string): Promise<void>
  badge(count: number): Promise<void>
  bounce(): Promise<void>
  focusWindow(): Promise<void>
}

/** The prompt that asks a person to admit an executable and the tree it
 *  launches, and returns their answer.
 *
 *  NOT one of the bindings above, and that is the whole distinction this seam
 *  exists to keep. Every binding is an effect only the native shell can
 *  perform — a picker, a banner, a badge, a window raise — so a client without
 *  a webview honestly has none of them. This is a prompt the RENDERER draws
 *  (agent-approval-prompt.tsx), which a plain browser draws exactly as well.
 *  Modelling it as a binding gave it a default that always rejected, and a
 *  browser-hosted client answered `unavailable` for it along with the six real
 *  ones — so the whole external-coordinator feature was unreachable outside a
 *  Wails build (nocx-qlp9w), and the always-rejecting default won the race the
 *  double mount created (nocx-pighx). */
export type ApprovalSurface = (facts: ApprovalFacts) => Promise<boolean>

/** What the person is being asked to admit, one fact per member. The wire
 *  carries three because the surface words them: a path and a digest glued
 *  into one string cannot be given a row each, and a durable scope key
 *  ("tool-endpoint:workspace:default") is an identifier this renderer must
 *  not parse to find a word for a person (nocx-fu18z). */
export interface ApprovalFacts {
  /** The agent's absolute path. */
  executable: string
  /** SHA-256 of that file's bytes. */
  digest: string
  /** The workspace the answer covers, by name. */
  workspace: string
}

/** The one binding name the reachability probe is asked about. All seven live
 *  on the same bound struct, so one answer covers the set: either this client
 *  is inside a Wails webview (or a shim that carries them) or it is not. */
const HOST_BINDING = 'main.WailsApp.HostFocusWindow'

/** The Wails event the shell emits when a person clicks a banner it
 *  presented. It carries the session id the banner was about. */
export const ATTENTION_ACTIVATED_EVENT = 'nocx:attentionActivated'

/** The Wails event seam. Narrow on purpose: this module needs exactly one
 *  subscription, and depending on the whole runtime would make it untestable
 *  outside a webview. */
export interface HostEvents {
  on(name: string, handler: (data: unknown) => void): () => void
}

const wailsBindings: HostBindings = {
  openFile: () => HostOpenFile(),
  openDirectory: () => HostOpenDirectory(),
  openUrl: (url) => HostOpenUrl(url),
  banner: (title, body, sessionId) => HostBanner(title, body, sessionId),
  badge: (count) => HostBadge(count),
  bounce: () => HostBounce(),
  focusWindow: () => HostFocusWindow(),
}

const wailsEvents: HostEvents = {
  on: (name, handler) => Events.On(name, (ev) => handler(ev)),
}

/**
 * Mount the client-host handler on the app's dispatcher. Returns the
 * unsubscribe function.
 *
 * bindings and events default to the real Wails runtime; both are injected so
 * the exchange can be exercised without one. A client with no reachable
 * bindings still mounts and still answers — failed, saying so — because the
 * coordinator must never be left waiting on a client that cannot act.
 */
export function mountClientHost(
  dispatcher: Dispatcher,
  bindings: HostBindings = wailsBindings,
  events: HostEvents = wailsEvents,
  approveAgent?: ApprovalSurface,
): () => void {
  const unsubscribeRequests = dispatcher.subscribe('host.request', (params) => {
    const p = params as HostRequest
    if (!p || !p.requestId || !p.capability) return
    void answer(dispatcher, bindings, approveAgent, p)
  })
  const unsubscribeEvents = events.on(ATTENTION_ACTIVATED_EVENT, (data) => {
    // The click half: the shell tells this renderer that a banner it
    // presented was activated, and the renderer tells the coordinator.
    // Nothing is done about it here — where the focus lands is the
    // coordinator's, because only it knows which connection holds the
    // session.
    const sessionId = activatedSessionId(data)
    if (!sessionId) return
    const activated: HostAttentionActivated = { sessionId }
    dispatcher.notify('host.attentionActivated', activated)
  })
  return () => {
    unsubscribeRequests()
    unsubscribeEvents()
  }
}

/** The Wails event payload is whatever the shell emitted. v3 wraps it in a
 *  WailsEvent whose `data` is the emitted value -- and a single emitted value
 *  may arrive either bare or as a one-element array, so both are read rather
 *  than one being assumed. Anything else is ignored: a click that cannot name
 *  a session is not a click this can honour. */
function activatedSessionId(data: unknown): string {
  const payload = (data as { data?: unknown } | null)?.data ?? data
  if (typeof payload === 'string') return payload
  if (Array.isArray(payload) && typeof payload[0] === 'string') return payload[0]
  return ''
}

/** What one capability produced: a picker's chosen path, a dismissal, or
 *  nothing at all for an effect that has no result. A typed outcome rather
 *  than a magic string, so 'no path' and 'the person cancelled' cannot be
 *  confused for one another by an empty value. */
interface Performed {
  path: string
  cancelled: boolean
  approved: boolean
}

async function answer(
  dispatcher: Dispatcher,
  bindings: HostBindings,
  approveAgent: ApprovalSurface | undefined,
  p: HostRequest,
): Promise<void> {
  // Can THIS CLIENT perform THIS capability — not "is this client native".
  // The six below are native effects and nothing else can produce them; the
  // seventh is a prompt this renderer draws, so what it needs is a mounted
  // surface and never a webview. One question, answered per capability,
  // because the two capabilities have genuinely different requirements
  // (nocx-qlp9w) — this is not a special case bolted onto a uniform rule.
  const unavailable =
    p.capability === 'agent.approval'
      ? approveAgent === undefined && 'this client has no agent approval surface'
      : !bindingReachable(HOST_BINDING) && 'this client has no native host'
  if (unavailable) {
    // A plain browser, the dev-web harness, the headless suite: there is no
    // shell here to open a picker or raise a banner. Said once, honestly, so
    // the coordinator answers its caller rather than waiting on a client that
    // will never act.
    //
    // UNAVAILABLE, NOT FAILED, and the two are different facts rather than
    // two words for one (nocx-bu8fl). `failed` is an effect that was
    // ATTEMPTED and did not happen — a denied permission, a thrown binding —
    // and the coordinator is right to remember it: a notification that was
    // accepted and never arrived earns a "Not delivered" row in the
    // notification centre. This client was never able to, and never will be,
    // which is the same fact as no client attached at all; the coordinator
    // maps it to its own no-UI-host answer, which notify exempts from that
    // feed because a channel that does not exist is not a channel that lost
    // a message. Answering `failed` here put a "Not delivered" row behind
    // every banner-routed notification in every browser-hosted client.
    resolve(dispatcher, { requestId: p.requestId, outcome: 'unavailable', error: unavailable })
    return
  }
  try {
    const done = await perform(bindings, approveAgent, p)
    if (done.cancelled) {
      resolve(dispatcher, { requestId: p.requestId, outcome: 'cancelled' })
      return
    }
    resolve(
      dispatcher,
      done.path
        ? { requestId: p.requestId, outcome: 'ok', path: done.path }
        : p.capability === 'agent.approval'
          ? { requestId: p.requestId, outcome: 'ok', approved: done.approved }
          : { requestId: p.requestId, outcome: 'ok' },
    )
  } catch (err) {
    const reason = err instanceof Error ? err.message : String(err)
    resolve(dispatcher, {
      requestId: p.requestId,
      outcome: 'failed',
      error: reason || 'the native host failed',
    })
  }
}

/** Perform one capability and say what it produced. */
async function perform(
  bindings: HostBindings,
  approveAgent: ApprovalSurface | undefined,
  p: HostRequest,
): Promise<Performed> {
  switch (p.capability) {
    case 'dialog.file':
      return picked(await bindings.openFile())
    case 'dialog.directory':
      return picked(await bindings.openDirectory())
    case 'shell.openUrl':
      await bindings.openUrl(p.url ?? '')
      return done
    case 'attention.banner':
      await bindings.banner(p.title ?? '', p.body ?? '', p.sessionId ?? '')
      return done
    case 'attention.badge':
      await bindings.badge(p.count ?? 0)
      return done
    case 'attention.bounce':
      await bindings.bounce()
      return done
    case 'window.focus':
      await bindings.focusWindow()
      return done
    case 'agent.approval':
      return {
        path: '',
        cancelled: false,
        // Non-null: answer() has already refused this capability when no
        // surface is mounted, which is the only way it can be absent here.
        approved: await approveAgent!({
          executable: p.executable ?? '',
          digest: p.digest ?? '',
          workspace: p.workspace ?? '',
        }),
      }
    default:
      // A capability this client does not know. The vocabulary is the
      // server's and closed, so this is a version skew — answered, never
      // dropped.
      throw new Error(`unknown host capability: ${String(p.capability)}`)
  }
}

/** The effect happened and produced nothing to report. */
const done: Performed = { path: '', cancelled: false, approved: false }

/** An empty path from a picker is a dismissal, which is the contract the
 *  Wails open dialog has always had. */
function picked(path: string): Performed {
  return path === ''
    ? { path: '', cancelled: true, approved: false }
    : { path, cancelled: false, approved: false }
}

function resolve(dispatcher: Dispatcher, params: HostResolved): void {
  dispatcher.call('host.resolved', params).catch(() => {
    // The broker refused the resolution -- a stale request id, because the
    // ask was dropped or its client died while the effect was in flight.
    // That is the server's honest answer to a request that is gone; nothing
    // to do here but stop.
    console.warn('nocx: host resolution refused (stale request?)')
  })
}

File: internal/app/clienthost/clienthost.go
// Package clienthost implements the native-host capabilities of a
// coordinator that has no window (nocx-uo1k6, design D3).
//
// Before the daemon split the backend WAS the Wails process, so the
// composition root injected Wails-backed implementations of
// transport.DialogService, transport.UrlOpener and notify.AttentionHost
// directly. The coordinator is now a process of its own with zero or more
// attached clients, and those three seams keep their meaning by being
// implemented HERE, against one narrow port: ask an attached client to
// perform the effect and report what it said.
//
// AD-3 read against this, because D3 gives the shell a job it did not have:
// the shell ends up IMPLEMENTING capabilities the coordinator asks for and
// owning no decision about them. Which pane a click focuses, which URL may
// be opened, whether a second picker may stack — every one of those stays on
// this side of the wire.
//
// Nothing here decides WHICH client serves an ask: ADR-0026 §16 decided that
// (broadcast, first-consumer-wins) and the broker implements it. The adapters
// call one method and map its outcome.
package clienthost

import (
	"context"
	"errors"
	"fmt"
	"log/slog"

	"github.com/shady2k/nocx/internal/notify"
	"github.com/shady2k/nocx/internal/transport"
)

// Requester is the coordinator's channel to an attached client: one ask, one
// answer, or an honest error. It is the whole of what this package depends
// on — the transport satisfies it, and a test satisfies it without a socket.
type Requester interface {
	RequestHost(ctx context.Context, ask transport.HostAsk) (transport.HostAnswer, error)
}

// Dialogs is transport.DialogService backed by an attached client.
//
// It deliberately does NOT implement transport.UploadPicker. That picker
// answers with a source ticket the renderer could not have authored (design
// R2), and the mint has to happen where the path is; with the backend in
// another process the picked path would have to cross the client to reach
// the mint, which is exactly the property R2 exists to deny. So
// dialog.openFileForUpload keeps reporting itself unavailable — the honest
// degrade — until that is designed rather than quietly weakened.
type Dialogs struct {
	req Requester
}

// NewDialogs builds the client-backed picker.
func NewDialogs(req Requester) *Dialogs { return &Dialogs{req: req} }

// OpenFile asks a client for the native file picker. The DialogService
// contract is unchanged and is now literally true of the far side too: the
// call returns when the person acts, "" means they cancelled, and nothing
// here can dismiss a picker that is already on screen — so the transport's
// capacity-one waiting gate is still what stops a second one stacking.
func (d *Dialogs) OpenFile(ctx context.Context) (string, error) {
	return d.pick(ctx, transport.HostCapOpenFile)
}

// OpenDirectory asks a client for the native directory picker. The file
// picker's sibling in every respect the coordinator can see.
func (d *Dialogs) OpenDirectory(ctx context.Context) (string, error) {
	return d.pick(ctx, transport.HostCapOpenDirectory)
}

func (d *Dialogs) pick(ctx context.Context, cap transport.HostCapability) (string, error) {
	answer, err := d.req.RequestHost(ctx, transport.HostAsk{Capability: cap})
	if err != nil {
		return "", err
	}
	if answer.Cancelled {
		// A dismissal is an OUTCOME, not a failure: the caller reads "" the
		// way it always has, and the transport turns it into a result
		// rather than an error (ws_dialog.go).
		return "", nil
	}
	return answer.Path, nil
}

// URLs is transport.UrlOpener backed by an attached client.
type URLs struct {
	req Requester
}

// NewURLs builds the client-backed browser opener.
func NewURLs(req Requester) *URLs { return &URLs{req: req} }

// OpenURL asks a client to open the URL in its platform browser. The
// transport has already refused anything that is not an http(s) URL with a
// host; this adapter adds no second gate, because a second answer to
// "may this be opened" is the duplication AGENTS.md names.
func (u *URLs) OpenURL(ctx context.Context, url string) error {
	_, err := u.req.RequestHost(ctx, transport.HostAsk{
		Capability: transport.HostCapOpenURL,
		URL:        url,
	})
	return err
}

// Attention is notify.AttentionHost backed by an attached client, plus the
// other half of a banner: what happens when a person clicks one.
type Attention struct {
	req   Requester
	log   *slog.Logger
	focus func(sessionID string)
}

// NewAttention builds the client-backed attention surface. focus is the
// composition root's session-focus push — the renderer owns session→tab and
// the backend cannot do it at all (nocx-wyp3p) — and may be nil on a host
// that has no such push, in which case a click only raises a window.
func NewAttention(req Requester, logger *slog.Logger, focus func(sessionID string)) *Attention {
	if logger == nil {
		logger = slog.Default()
	}
	return &Attention{req: req, log: logger, focus: focus}
}

// Banner asks a client to present one desktop banner. Only the event's
// presentation fields and its addressing session id cross the wire — the
// same three the Wails adapter reads (internal/notify/wailsadapter).
func (a *Attention) Banner(ctx context.Context, ev notify.Event) error {
	_, err := a.req.RequestHost(ctx, transport.HostAsk{
		Capability: transport.HostCapBanner,
		Title:      ev.Title,
		Body:       ev.Body,
		SessionID:  ev.SessionID,
	})
	return unavailableIfNoHost(err)
}

// Badge asks a client to set its dock badge; 0 clears it.
func (a *Attention) Badge(ctx context.Context, count int) error {
	_, err := a.req.RequestHost(ctx, transport.HostAsk{
		Capability: transport.HostCapBadge,
		Count:      count,
	})
	return unavailableIfNoHost(err)
}

// Bounce asks a client for the attention bounce.
func (a *Attention) Bounce(ctx context.Context) error {
	_, err := a.req.RequestHost(ctx, transport.HostAsk{Capability: transport.HostCapBounce})
	return unavailableIfNoHost(err)
}

// Activated is transport.AttentionActivation: a person clicked a banner a
// client presented. Two halves, and neither is the client's to decide.
//
// The WINDOW is raised by asking a client, because only a shell has a window
// manager. The PANE is focused by the coordinator's own session-focus push,
// because the renderer owns session→tab and the destination is resolved from
// the session's current subscriber — which may be a different connection
// from the one that showed the banner. A click on a session no pane holds
// moves the window and nothing else, which is the honest outcome rather than
// an error (unchanged from before the split).
func (a *Attention) Activated(ctx context.Context, sessionID string) {
	if _, err := a.req.RequestHost(ctx, transport.HostAsk{Capability: transport.HostCapFocusWindow}); err != nil {
		// Said out loud and not fatal: the pane focus below is still worth
		// attempting, and a click that moves the tab without raising the
		// window is better than one that does nothing.
		a.log.Warn("notification click could not raise a window", "sessionId", sessionID, "error", err)
	}
	if a.focus == nil {
		a.log.Error("notification click cannot be honored: no session-focus path is wired", "sessionId", sessionID)
		return
	}
	a.focus(sessionID)
}

// unavailableIfNoHost reports a missing client as notify.ErrUnavailable AS
// WELL AS the capability's own error.
//
// Both words are true and both are load-bearing. "No UI host attached" is
// what a caller asked about; "unavailable on this host" is what the notify
// pipeline's one exemption from the failure feed is written against — a
// channel that does not exist right now is not a channel that lost a
// message. Wrapping both keeps that exemption working without a second
// spelling of absence (AD-8), and errors.Is answers either question.
func unavailableIfNoHost(err error) error {
	if err == nil {
		return nil
	}
	if errors.Is(err, transport.ErrNoUIHost) {
		return fmt.Errorf("%w: %w", notify.ErrUnavailable, err)
	}
	return err
}
